PYME/recipes/kinetics_recipes.py

```diff
@@ -0,0 +1,428 @@
+
+        start_times = ds['t']
+        stop_times = ds['t2']
+
+        on_times = np.zeros_like(ds['t'])
+        off_times = np.zeros_like(ds['t'])
+
+        for i in np.unique(ds[self.labelKey]):
+
+            clump_data_idxs = np.where(ds[self.labelKey] == i)[0]
+
+            on_times[clump_data_idxs] = stop_times[clump_data_idxs] - start_times[clump_data_idxs] + 1
+
+            shifted_start_times = np.roll(start_times[clump_data_idxs], -1)
+
+            #TODO: this approach for the final off time might be wrong. OR would have to treat it differently as bleaached
+            cluster_stop_times = stop_times[clump_data_idxs]
+
+            if self.keepEndBlinks == True:
+                shifted_start_times[-1] = last_frame
+            elif self.keepEndBlinks == False:
+                shifted_start_times[-1] = cluster_stop_times[-1] + 1
+
+            loop_offtimes = shifted_start_times - cluster_stop_times - 1
+
+            off_times[clump_data_idxs] = np.round(loop_offtimes).astype('float')
+
+        try:
+            mdh = namespace[self.inputName].mdh        
+            ds.mdh = mdh
+            ds.mdh.setEntry('bad_offtimes.number_blinks_lost', len(np.where(off_times <= 0)[0]))
+            ds.mdh.setEntry('bad_offtimes.blinks_remaining', len(np.where(off_times > 0)[0]))
+        except:
+            pass
+
+        #todo: filter out bad points here, no point in blowing out non-optional filtering into unique modules, the whole thing will be cleaner then
+
+
+        ds.addColumn(self.onTimesColName, on_times)
+        ds.addColumn(self.offTimesColName, off_times)
+
+        namespace[self.outputName] = ds
+
+
+@register_module('FitSwitchingRates')
+class FitSwitchingRates(ModuleBase):
+    """
+    Calculates switching rates based on data which has been run through a blink finding script (trackutils.findtracks2),
+    a clustering script (hdbscan), and a script which calculates all unique on and off states (findblinkstatedurations)
+    """
+    inputName = Input('kinetics')
+    outputName = Output('kinetics_out')
+    outputEmphistData = Output('emphist_data')
+    onTimesColName = CStr('on_times')
+    offTimesColName = CStr('off_times')
+
+    offTimesBinSizeInSeconds = Float(2.5)
+
+    minOffTimeSecondsToFit = Float(0.0)
+    maxOffTimeSecondsToFit = Float(5000.0)
+
+    off_init_condition_multiple = Int(1)
+
+    exposure_time = Float(-1.)
+
+    def execute(self, namespace):
+        from PYME.Analysis.BleachProfile.kinModels import eimod
+
+        #todo: add higher order fits
+
+        oncol = self.onTimesColName
+        offcol = self.offTimesColName
+
+        inp = namespace[self.inputName]
+
+        if 'mdh' in dir(inp):
+            mdh = inp.mdh
+            cyctime = mdh.getEntry('Camera.CycleTime')
+        else:
+            cyctime = self.exposure_time
+
+        ds = tabular.mappingFilter(inp)
+
+        off_fit_limit_index = np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int')
```

Comment:
your bin setup / rounding, +1 and then [:-1] trimming is a bit odd and without really following through I'm worried that you might sometimes be slightly off from what you intend (e.g. when you use `bins_on[:1]` in the fit, do you actually want those bin edges, or would you rather use the bin centers?). Maybe have a look at whether this can be simplified, and if not include some comments to make it easier to follow what's going on?

Reply:
I do want those bin edges specifically for the fit, since its an integrated fit for histograms. I'll add comments to explain.